# Convolutional Neural Networks

## Practical 1

_**Building a Convolutional Neural Network (CNN) from Scratch on MNIST dataset.**_

NOTE: This notebook is recommended to run on GPU to save time during model training.

In [22]:
# Imports required packages

import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import mnist

In [8]:
# Loads MNIST dataset
# NOTE: Downloading for the first time may take few minutes to complete

mnist = tf.keras.datasets.mnist.load_data()

11490434/11490434 [==============================] - 3s 0us/step


In [10]:
# Considering dataset is organized in tuple, items are referenced as follows
(X_train_full, y_train_full), (X_test, y_test) = mnist

In [12]:
# Checks the shape of the datasets

print("Full training set shape:", X_train_full.shape)
print("Test set shape:", X_test.shape)

Full training set shape: (60000, 28, 28)
Test set shape: (10000, 28, 28)


In [14]:
# Normalizes the data between 0 and 1 for effective neural network model training
X_train_full = X_train_full / 255.
X_test = X_test / 255.

In [16]:
# Splits train dataset further to seperate 5000 instances to be used as validation set

X_train, X_val = X_train_full[:-5000], X_train_full[-5000:]
y_train, y_val = y_train_full[:-5000], y_train_full[-5000:]

In [24]:
# To match the input shape of the CNN model, a channel dimention gets added to each dataset

X_train = X_train[..., np.newaxis]
X_val = X_val[..., np.newaxis]
X_test = X_test[..., np.newaxis]

In [26]:
# Checks for the updated shape
X_train.shape

(55000, 28, 28, 1)

In [28]:
# Creates CNN model by having convoluted, pooling, dropout and dense layer in the specified order for this experiment.
# Each convoluted layer is further initialized with specific kernel size, padding, activation and initialization.

tf.random.set_seed(42)

model = tf.keras.Sequential([
    tf.keras.layers.Conv2D(32, kernel_size=3, padding="same", activation="relu", kernel_initializer="he_normal"),
    tf.keras.layers.Conv2D(64, kernel_size=3, padding="same", activation="relu", kernel_initializer="he_normal"),
    tf.keras.layers.MaxPool2D(),
    tf.keras.layers.Flatten(),
    tf.keras.layers.Dropout(0.25),
    tf.keras.layers.Dense(128, activation="relu", kernel_initializer="he_normal"),
    tf.keras.layers.Dropout(0.5),
    tf.keras.layers.Dense(10, activation="softmax")
])

model.compile(loss="sparse_categorical_crossentropy", optimizer="nadam", metrics=["accuracy"])

2024-08-28 10:29:00.610493: I tensorflow/core/common_runtime/process_util.cc:146] Creating new thread pool with default inter op setting: 2. Tune using inter_op_parallelism_threads for best performance.


In [30]:
# Fits the model.
model.fit(X_train, y_train, epochs=10, validation_data=(X_val, y_val))

Epoch 1/10
1719/1719 [==============================] - 516s 300ms/step - loss: 0.1911 - accuracy: 0.9420 - val_loss: 0.0433 - val_accuracy: 0.9878
Epoch 2/10
1719/1719 [==============================] - 514s 299ms/step - loss: 0.0795 - accuracy: 0.9751 - val_loss: 0.0449 - val_accuracy: 0.9868
Epoch 3/10
1719/1719 [==============================] - 512s 298ms/step - loss: 0.0601 - accuracy: 0.9819 - val_loss: 0.0329 - val_accuracy: 0.9922
Epoch 4/10
1719/1719 [==============================] - 493s 287ms/step - loss: 0.0477 - accuracy: 0.9851 - val_loss: 0.0364 - val_accuracy: 0.9902
Epoch 5/10
1719/1719 [==============================] - 487s 283ms/step - loss: 0.0408 - accuracy: 0.9871 - val_loss: 0.0325 - val_accuracy: 0.9916
Epoch 6/10
1719/1719 [==============================] - 478s 278ms/step - loss: 0.0344 - accuracy: 0.9892 - val_loss: 0.0450 - val_accuracy: 0.9906
Epoch 7/10
1719/1719 [==============================] - 511s 297ms/step - loss: 0.0297 - accuracy: 0.9907 - val_

In [49]:
# Saves the trained model for later reference
# NOTE: Make sure the folder "models" exists under the current working directory

model.save("./models/my_mnist_cnn_model.keras")

2024-08-28 11:53:42.973180: I tensorflow/core/common_runtime/executor.cc:1197] [/device:CPU:0] (DEBUG INFO) Executor start aborting (this does not indicate an error and you can ignore this message): INVALID_ARGUMENT: You must feed a value for placeholder tensor 'inputs' with dtype float and shape [?,12544]
	 [[{{node inputs}}]]
2024-08-28 11:53:42.980319: I tensorflow/core/common_runtime/executor.cc:1197] [/device:CPU:0] (DEBUG INFO) Executor start aborting (this does not indicate an error and you can ignore this message): INVALID_ARGUMENT: You must feed a value for placeholder tensor 'inputs' with dtype float and shape [?,128]
	 [[{{node inputs}}]]
2024-08-28 11:53:43.100294: I tensorflow/core/common_runtime/executor.cc:1197] [/device:CPU:0] (DEBUG INFO) Executor start aborting (this does not indicate an error and you can ignore this message): INVALID_ARGUMENT: You must feed a value for placeholder tensor 'inputs' with dtype float and shape [?,12544]
	 [[{{node inputs}}]]
2024-08-28 1

INFO:tensorflow:Assets written to: ./models/my_mnist_cnn_model/assets


INFO:tensorflow:Assets written to: ./models/my_mnist_cnn_model/assets


In [51]:
# Evaluates the model on test dataset
model.evaluate(X_test, y_test)

313/313 [==============================] - 6s 18ms/step - loss: 0.0245 - accuracy: 0.9933


[0.024464121088385582, 0.9933000206947327]

Accuracy of the model is observed to be 99.33% on the test dataset.